### DISCUSSION TITLE
IT regression area Folder structure

### DISCUSSION THREAD

**Discussion by MANISH-K-07** [checkstyle/checkstyle#14757] [General]
From https://github.com/checkstyle/checkstyle/issues/6981#issue-481074941 : 

> Our regression area in IT should be fixed to match our pattern in Test folder.
> Google and Sun area should atleast start with Input, but their folder structure follows a different pattern.

Now that we have completed decorating more than half of the input files of regression area in IT folder (the rest will also be done soon),
do we also want the folder structure to match our test path?

We have the test path folders organized and grouped into 2 levels 
(coding, annotation, naming, blocks, etc as first level with respective sub-folders).

We donot have as many files in IT but IMO, implementing a folder hierarchy would certainly help with organizing files at https://github.com/checkstyle/checkstyle/tree/master/src/it/java/org/checkstyle/suppressionxpathfilter

@rnveach , this is an extension of your findings and please share your opinions if we need it done @romani , @nrmancuso

**Comment by rnveach** [comment]
I am more leaning to that I don't feel Google/Sun should change their folder pattern. It correlates to their style document , and I think it makes it easier to point to those areas and ensure we have tests covering what we support.

I don't know what pattern the others should follow right now. I leave that for others to bring ideas.

**Reply by MANISH-K-07** [reply]
> regression area in IT folder

Not Google or Sun, just our xpath folders

> I don't know what pattern the others should follow right now.

@rnveach , why not the same as our test path?

**Reply by rnveach** [reply]
There is also imports, but it is small compared to xpath.

If we want to get fully technical, our test area names the input after the test being conducted, and the test class is named after the production class being tested.

`XpathRegression` is basically testing against 2 things, `SuppressionXpathFilter` and the check. Originally some of our checks didn't support xpath suppression because they were logging on a line/column number and not the AST. The AST is needed for xpath suppression. Even before that, xpath suppression had some corner cases where it couldn't suppress specific areas because of the xpath limitation (Not sure, but there may still be areas like this. I had an issue on it a while ago).

For us to align the IT area, we would need to know which of the 2 we are testing against the most to base the names off of. If it is `SuppressionXpathFilter` then every check should be in 1 test class and the inputs would need further renames to align to `InputSuppressionXpathFilter...`. If it is the check, then everything should be renamed to the check name and the inputs would again need to be renamed to `Input{CheckName}`.

If we just want to have the input name and folders be based off the test name and not change the existing test names, then I am fine with that. However, packages in the inputs would need to be updated when moving to the new folder.

Truthfully any reworking may possibly require to redo the PRs you have done, or start new ones.

**Reply by romani** [reply]
We can do extra folders at 
https://github.com/checkstyle/checkstyle/tree/master/src/it/java/org/checkstyle/suppressionxpathfilter to group test in same way as Checks are grouped.

No other changes, other ITs should not be changed. We have special project that describes how them should be organized.

**Reply by MANISH-K-07** [reply]
> We can do extra folders at
> https://github.com/checkstyle/checkstyle/tree/master/src/it/java/org/checkstyle/suppressionxpathfilter to group test in same way as Checks are grouped.
----
> We donot have as many files in IT but IMO, implementing a folder hierarchy would certainly help with organizing files at https://github.com/checkstyle/checkstyle/tree/master/src/it/java/org/checkstyle/suppressionxpathfilter

Yes @romani , precisely my intention of this discussion.

>  We have special project that describes how them should be organized.

Just curious, what project are we referring to? Refine google style guide for Gsoc?

**Reply by MANISH-K-07** [reply]
>  If it is the check, then everything should be renamed to the check name and the inputs would again need to be renamed to Input{CheckName}

> If we just want to have the input name and folders be based off the test name and not change the existing test names, then I am fine with that.

@rnveach , IMO the distinction "Xpath" in the names would be helpful.
So, maybe we could do the second point you suggested.. grouping based on test name, just to overcome the way our xpath tests are dispersed in first order right now.

**Reply by romani** [reply]
>Just curious, what project are we referring to? Refine google style guide for Gsoc?

Yes

**Reply by rnveach** [reply]
Not sure why xpath IT is connected to Google Style. They are not related. If the GSOC wiki doesn't specifically mentioned this, then it needs to be updated.

**Reply by MANISH-K-07** [reply]
Not xpath @rnveach , we can update xpath files now.
By gsoc project, @romani was referring to the Google folders, I believe.
Please see https://github.com/checkstyle/checkstyle/discussions/14757#discussioncomment-9003723

**Comment by nrmancuso** [comment]
> please share your opinions if we need it done

I don’t see any value in this work personally, so my answer is “no”.

**Reply by MANISH-K-07** [reply]
@nrmancuso , I agree that this doesn't make any "worthy" difference from user's pov.
The one benefit however, from grouping the xpath files/folders could be the ease of search for dev team/contributors, I believe.

As of now, all the xpath tests are dispersed in first order and finding the targeted file is a bit confusing without search filter
(I experienced this first hand as part of #14715). Please see https://github.com/checkstyle/checkstyle/tree/master/src/it/java/org/checkstyle/suppressionxpathfilter

The update is rather simple and maybe beneficial for this specific path of our codebase ?

**Comment by MANISH-K-07** [comment]
@romani , @rnveach , 
@nrmancuso , please confirm if we want this discussion transferred to an issue or not?

**Reply by romani** [reply]
I amok  to  make issue, and suggest to move all of same package in single PR, so it will naturally be split in few PRs.

**Reply by MANISH-K-07** [reply]
Sure, Please see #14769
